src/Plutarch/Test/QuickCheck.hs

```diff
@@ -77,120 +84,226 @@ import Test.QuickCheck (
+type family IsLast (h :: Type) :: Bool where
+    IsLast (_ -> _) = 'False
+    IsLast _ = 'True
+
+-- Helper for type error.
+-- Empty constraint when all arguments/return are TestableTerm
+-- Type error when there exists some non TestableTerm type.
+type family OnlyTestableTerm (h :: Type) :: Constraint where
+    OnlyTestableTerm (TestableTerm _ -> a) = OnlyTestableTerm a
+    OnlyTestableTerm (TestableTerm _) = ()
+    OnlyTestableTerm h =
+        TypeError
+            ( 'Text "\""
+                ':<>: 'ShowType h
+                ':<>: 'Text "\" is not in terms of \"TestableTerm\""
+                ':$$: 'Text "\tFunction should be only in terms of TestableTerms"
+            )
+
+{- | Wraps TestableTerm lambda to `PFun` for function generation and shrinking.
+
+ @since 2.1.0
+-}
+class PWrapLam' (h :: Type) (lam :: Bool) (last :: Bool) where
+    pwrapLam' :: h -> PLamWrapped h
 
--- | @since 2.0.0
-type family PTT' (isplam :: Bool) (end :: Bool) :: PTermType where
-    PTT' 'True 'True = 'LastPFunction
-    PTT' 'False 'True = 'LastPTerm
-    PTT' 'True 'False = 'PFunction
-    PTT' 'False 'False = 'PTerm
+instance
+    forall (a :: Type) (b :: Type) (pa :: S -> Type) (pb :: S -> Type).
+    ( TestableTerm (pa :--> pb) ~ a
+    , PWrapLam' b (IsLam b) (IsLast b)
+    , PLamWrapped (a -> b) ~ (PFun pa pb -> PLamWrapped b)
+    , PLift pa
+    , PLift pb
+    ) =>
+    PWrapLam' (a -> b) 'True 'False
+    where
+    pwrapLam' f (PFn pf) = pwrapLam' @b @(IsLam b) @(IsLast b) $ f (TestableTerm pf)
 
--- | @since 2.0.0
-type PTT (p :: S -> Type) = PTT' (IsPLam p) (IsLast p)
+instance
+    forall (a :: Type) (b :: Type).
+    ( PWrapLam' b (IsLam b) (IsLast b)
+    , PLamWrapped (a -> b) ~ (a -> PLamWrapped b)
+    ) =>
+    PWrapLam' (a -> b) 'False 'False
+    where
+    pwrapLam' f x = pwrapLam' @b @(IsLam b) @(IsLast b) $ f x
 
--- | @since 2.0.0
-type family IsPLam (p :: S -> Type) :: Bool where
-    IsPLam ((a :--> b) :--> c) = 'True
-    IsPLam _ = 'False
+instance
+    forall (a :: Type).
+    (PLamWrapped a ~ a) =>
+    PWrapLam' a 'False 'True
+    where
+    pwrapLam' = id
 
--- | @since 2.0.0
-type family IsLast (p :: S -> Type) :: Bool where
-    IsLast (a :--> PBool) = 'True
-    IsLast _ = 'False
+{- | Constraint for `PWrapLam'` that will give a better type error message.
+
+ @since 2.1.0
+-}
+type PWrapLam (h :: Type) = (PWrapLam' h (IsLam h) (IsLast h), OnlyTestableTerm h)
 
-{- | Finds Haskell level TestableTerm equivlance of Plutarch
-     functions. This TypeFamily expects the input Plutarch functions to
-     be returning @PBool@ at the end.
+{- | Replace any TestableTerm that is a lambda with `PFun` for generation and
+     shrinking. It will ignore lambda on the return as it should not be
+     generated.
 
-     This is used to find type signatures for @quickCheck@able
-     functions from Plutarch terms like @Term s (a :--> b :--> PBool)@.
+= Note
+"TestableTerm that is lambda" means @TestableTerm (a :--> b)@.
```

Comment:
'Lambda' tends to imply anonymity. If we're just talking about things of type `a :--> b`, those are 'Plutarch functions'.

Reply:
Also, you probably want `'TestableTerm'` so we get Haddock links.